### PR TITLE
feat: send only patch updates & split up PK changes or PK-less updates

### LINKED ISSUE #165: Send only changes when something is updated
> One change we can implement immediately however is moving to a patch update model. Currently everything assumes that each message contains the entire value of each row. This is not ideal as it means updates to shapes with large rows will be pretty heavy to send to clients. It's much more efficient and fast to send just the update. So we'll want to update the `Shape` TS class to merge in updates as well as the Redis & Bash clients (or perhaps just delete the bash client as it's more for fun than serious — bash is annoying to write).

_Originally posted by @KyleAMathews in https://github.com/electric-sql/electric-next/issues/61#issuecomment-2243595052_

For updates, store & send only the changed fields + a PK, and for deletes store & send only the PK. PK is enough for us to reason about the change.

### PR DESCRIPTION
**Pull request by icehaunter**
Closes #165, #181, #170.

This includes changes to the `Shape` class in JS to apply patches correctly.

### CODE REVIEW

**Comment by balegas** [comment]
hold merging this one before we make a decision on keying because of splitting on pk changes.

**Review by kevin-dp — APPROVED** [review]
Brilliant! No comments, really 🔥

File: packages/sync-service/test/electric/plug/router_test.exs [hunk 1]
```diff
@@ -210,5 +210,157 @@ defmodule Electric.Plug.RouterTest do
...
+        end)
+
+      Postgrex.query!(db_conn, "UPDATE wide_table SET value2 = 'test value 2' WHERE id = 1", [])
+
+      assert %{status: 200} = conn = Task.await(task)
+
+      assert [%{"key" => ^key, "value" => value}, _] = Jason.decode!(conn.resp_body)
+
+      # No extra keys should be present, so this is equality assertion, not a match
+      assert value == %{"id" => "1", "value2" => "test value 2"}
```
Comment (alco) [inline packages/sync-service/test/electric/plug/router_test.exs:245]: You can get equivalent behaviour (no extra keys allowed) by matching on the variable. I think it reads nicer this way:
```suggestion
      value == %{"id" => "1", "value2" => "test value 2"}
      assert [%{"key" => ^key, "value" => ^value}, _] = Jason.decode!(conn.resp_body)
```

**Review by alco — APPROVED** [review]
:trophy:

**Pull request merged**